aten/src/ATen/cpu/vec/vec512/vec512_base.h

```diff
@@ -0,0 +1,869 @@
+#pragma once
+
+// DO NOT DEFINE STATIC DATA IN THIS HEADER!
+// See Note [Do not compile initializers with AVX]
+
+#include <cstring>
+#include <functional>
+#include <cmath>
+#include <type_traits>
+#include <bitset>
+
+#include <ATen/cpu/vec/vec512/intrinsics.h>
+#include <ATen/native/Math.h>
+#include <ATen/NumericUtils.h>
+#include <c10/util/C++17.h>
+#include <c10/util/BFloat16.h>
+#include <c10/util/BFloat16-math.h>
+#include <c10/util/copysign.h>
+#include <c10/util/math_compat.h>
+#include <ATen/native/cpu/zmath.h>
+#include <c10/util/TypeCast.h>
+#include <c10/macros/Macros.h>
+
+#if defined(__GNUC__)
+#define __at_align64__ __attribute__((aligned(64)))
+#elif defined(_WIN32)
+#define __at_align64__ __declspec(align(64))
+#else
+#define __at_align64__
+#endif
+
+namespace at {
+namespace vec {
+// See Note [Acceptable use of anonymous namespace in header]
+namespace {
+// at::Half and at::BFloat16 should be treated as floating point
+template <typename T>
+struct is_floating_point:
+    std::integral_constant<bool,
+      std::is_floating_point<T>::value ||
+      std::is_same<T, at::Half>::value ||
+      std::is_same<T, at::BFloat16>::value> {
+};
+
+template<size_t n> struct int_of_size;
+
+#define DEFINE_INT_OF_SIZE(int_t) \
+template<> struct int_of_size<sizeof(int_t)> { using type = int_t; }
+
+DEFINE_INT_OF_SIZE(int64_t);
+DEFINE_INT_OF_SIZE(int32_t);
+DEFINE_INT_OF_SIZE(int16_t);
+DEFINE_INT_OF_SIZE(int8_t);
+
+#undef DEFINE_INT_OF_SIZE
+
+template <typename T>
+using int_same_size_t = typename int_of_size<sizeof(T)>::type;
+
+// NOTE: If you specialize on a type, you must define all operations!
+
+// emulates Vectorizedtorized types
+template <class T>
+struct Vectorized {
+private:
+  __at_align64__ T values[64 / sizeof(T)];
+public:
+  using value_type = T;
+  using size_type = int;
+  // See Note [constexpr static function to avoid odr-usage compiler bug]
+  static constexpr size_type size() {
+    return 64 / sizeof(T);
+  }
+  Vectorized() : values{0} {}
+  Vectorized(T val) {
+    for (int i = 0; i != size(); i++) {
+      values[i] = val;
+    }
+  }
+  template<typename... Args,
+           typename = std::enable_if_t<(sizeof...(Args) == size())>>
+  Vectorized(Args... vals) : values{vals...}{
+  }
+  // This also implies const T& operator[](int idx) const
+  inline operator const T*() const {
+    return values;
+  }
+  // This also implies T& operator[](int idx)
+  inline operator T*() {
+    return values;
+  }
+  template <int64_t mask_>
+  static Vectorized<T> blend(const Vectorized<T>& a, const Vectorized<T>& b) {
+    int64_t mask = mask_;
+    Vectorized Vectorized;

```

Comment:
you really want to name this local Vectorized? ;)

Reply:
Uh, oh! I guess I should've done such sanity checks after running a string substitution script.

Reply:
this happens a lot below too

Reply:
Thanks, I'll change those variable names tonight.